File: sdk/java/src/main/java/com/anyvali/ParseResult.java
package com.anyvali;

import java.util.List;

/**
 * Result of a safeParse() call.
 */
public record ParseResult<T>(
        boolean success,
        T data,
        List<ValidationIssue> issues
) {
    public ParseResult {
        issues = issues != null ? List.copyOf(issues) : List.of();
    }

    public static <T> ParseResult<T> success(T data) {
        return new ParseResult<>(true, data, List.of());
    }

    public static <T> ParseResult<T> failure(List<ValidationIssue> issues) {
        return new ParseResult<>(false, null, issues);
    }
}

File: sdk/kotlin/src/main/kotlin/com/anyvali/ParseResult.kt
package com.anyvali

sealed class ParseResult<out T> {
    data class Success<T>(val value: T) : ParseResult<T>()
    data class Failure(val issues: List<ValidationIssue>) : ParseResult<Nothing>()

    val isSuccess: Boolean get() = this is Success
    val isFailure: Boolean get() = this is Failure

    fun getOrNull(): T? = when (this) {
        is Success -> value
        is Failure -> null
    }

    fun getOrThrow(): T = when (this) {
        is Success -> value
        is Failure -> throw ValidationError(issues)
    }

    fun issuesOrEmpty(): List<ValidationIssue> = when (this) {
        is Success -> emptyList()
        is Failure -> issues
    }
}
